lib/plugins/aws/deploy/lib/uploadArtifacts.js

```diff
@@ -76,36 +77,36 @@ module.exports = {
   },
 
   uploadFunctions() {
-    let shouldUploadService = false;
     this.serverless.cli.log('Uploading artifacts...');
+
     const functionNames = this.serverless.service.getAllFunctions();
-    return BbPromise.map(functionNames, (name) => {
-      const functionArtifactFileName = this.provider.naming.getFunctionArtifactName(name);
-      const functionObject = this.serverless.service.getFunction(name);
-      functionObject.package = functionObject.package || {};
-      let artifactFilePath = functionObject.package.artifact ||
-        this.serverless.service.package.artifact;
-      if (!artifactFilePath ||
-        (this.serverless.service.artifact && !functionObject.package.artifact)) {
-        if (this.serverless.service.package.individually || functionObject.package.individually) {
-          const artifactFileName = functionArtifactFileName;
-          artifactFilePath = path.join(this.packagePath, artifactFileName);
-          return this.uploadZipFile(artifactFilePath);
+    const artifactFilePaths = _.uniq(
+      _.map(functionNames, (name) => {
+        const functionArtifactFileName = this.provider.naming.getFunctionArtifactName(name);
+        const functionObject = this.serverless.service.getFunction(name);
+        functionObject.package = functionObject.package || {};
+        const artifactFilePath = functionObject.package.artifact ||
+          this.serverless.service.package.artifact;
+
+        if (!artifactFilePath ||
+          (this.serverless.service.artifact && !functionObject.package.artifact)) {
+          if (this.serverless.service.package.individually || functionObject.package.individually) {
+            const artifactFileName = functionArtifactFileName;
+            return path.join(this.packagePath, artifactFileName);
+          }
+          return this.provider.naming.getServiceArtifactName();
         }
-        shouldUploadService = true;
-        return BbPromise.resolve();
-      }
+
+        return artifactFilePath;
+      })
+    );
+
+    return BbPromise.map(artifactFilePaths, (artifactFilePath) => {
+      const stats = fs.statSync(artifactFilePath);
+      this.serverless.cli.log(`Uploading service .zip file to S3 (${filesize(stats.size)})...`);
       return this.uploadZipFile(artifactFilePath);
-    }, { concurrency: 3 }).then(() => {
-      if (shouldUploadService) {
-        const artifactFileName = this.provider.naming.getServiceArtifactName();
-        const artifactFilePath = path.join(this.packagePath, artifactFileName);
-        const stats = fs.statSync(artifactFilePath);
-        this.serverless.cli.log(`Uploading service .zip file to S3 (${filesize(stats.size)})...`);
-        return this.uploadZipFile(artifactFilePath);
-      }
-      return BbPromise.resolve();
-    });
+    }, { concurrency: 3 }
```

Comment:
We could define this as constant above in the file (e.g. `NUM_CONCURRENT_UPLOADS: 3`) and use that here instead of the literal number. 
Additionally, it might be useful to have an override in `serverless.yml` under the provider property to change the number if needed. Of course, if 3 is ok for all uses, we do not need that.

Reply:
@HyperBrain, I agree about the constant and did it.

About the option to override in `serverless.yml` IMO we should move on with this constant and check if users will want/need it in the future.

Reply:
Yep. Agree. If there is need for it, we'll see it.